sympy/solvers/solveset.py

```diff
@@ -1961,7 +2020,7 @@ def add_type(lhs, rhs, symbol, domain):
         assert (len(rhs_s.args)) == 1
         rhs = rhs_s.args[0]
 
-        if lhs.is_Add:
+        if (lhs - rhs).is_Add:
             result = add_type(lhs, rhs, symbol, domain)
```

Comment:
Why is this change needed?

Reply:
If suppose lhs have only 1 term and rhs is some finite set then I think it would also be a add type... That is how _is_lambert case has been removed from here.

Reply:
I'm sorry but I don't understand what you mean

Reply:
for e.g.:

```
>>> solveset(x*exp(x) - 1, x,Reals)
INTERMEDIATE VALUES...
_R*exp(_R)  ---> lhs 
# if it L-2023 was not modified then will not be considered as `is_Add` (which will ultimately returns `conditionset`) 

 _R*exp(_R) - 1 ---> lhs - rhs
# but as of now ( after modification) this will be considered as `is_Add`


FiniteSet(LambertW(1))
```
The other way is not to disturb the `_is_lambert` function below this line. (As of now it is removed) .

Hope I've cleared the point...

Reply:
I don't think I understand what the purpose of this function or this line of code is at all.

Reply:
I think the intent of making such a statement is to validate between `Add` type or `Mul` type(of `lhs`)... 
But as of now, I think the purpose is not necessary.(Since without that statement the same purpose is achieved.) So, if statement should be removed as of now.
May in future there may be the need for `Mul` type extension, So IMHO not to remove the function `add_type`.

Reply:
> I think the intent of making such a statement is to validate between `Add` type or `Mul` type

But why? I don't understand what role this function has in `solveset`.

Reply:
> But why? I don't understand what role this function has in solveset.

I too, don't have the reason, since I haven't written that code. That is already present in code base.

https://github.com/sympy/sympy/blob/b564d9ba2705ee2978766e1ee2102750834df68b/sympy/solvers/solveset.py#L1936-L1969

Reply:
> If suppose `lhs` have only 1 term and `rhs` is some entity(like an integer ) then I think it would also be a `is_Add` type... That is how `_is_lambert` case has been removed from here.

I think that is not an "add type" equation and should possibly be handled by `invert_complex`.

Reply:
> > If suppose lhs have only 1 term and rhs is some entity(like an integer ) then I think it would also be a is_Add type... That is how _is_lambert case has been removed from here.
> 
> I think that is not an "add type" equation and should possibly be handled by invert_complex.

Which one?

Reply:
The docstring says equations of the form `a*f(x) + b*g(x) + .... = c` but you said the lhs only has one term and the rhs is an integer so it's something like `f(x) = c`. In that case we should try to invert the equation rather than use the "add type" handler.

Reply:
> The docstring says equations of the form a*f(x) + b*g(x) + .... = c but you said the lhs only has one term and the rhs is an integer so it's something like f(x) = c. In that case we should try to invert the equation rather than use the "add type" handler.

I think that's the reason they had again a separate case of `_is_lambert` again after `add_type` since sometimes invert can't handle Lambert solutions smoothly (as in the case of https://github.com/sympy/sympy/pull/21025#discussion_r589207626)

Reply:
I think it would be better to handle this in invert if possible. I guess you mean an equation like `x*exp(x) = y`. That can be handled in `invert_real` under the `Mul` section.

Reply:
> I think it would be better to handle this in invert if possible. I guess you mean an equation like x*exp(x) = y. That can be handled in invert_real under the Mul section.

There are many forms of representation and many different ways of solving lambert equations, I think it's not necessary to rewrite it again in `invert` function as it's already in `_solve_lambert`. (also see [different types of lambert representation](https://math.stackexchange.com/questions/1048034/how-to-check-if-some-equation-can-be-solved-using-lambert-operatornamew-fun)  and https://github.com/sympy/sympy/issues/20545#issuecomment-739461506)
https://github.com/sympy/sympy/blob/b564d9ba2705ee2978766e1ee2102750834df68b/sympy/solvers/bivariate.py#L194

Reply:
Please do reply to this comment so that I could proceed for probably a final commit for this PR.

Reply:
I'm still waiting for you to explain why this line of code is changed. It was clearly written as `lhs.is_Add` on purpose. The change to `(lhs - rhs).is_Add` means that it is almost always `True`.

To me it looks like if there is a reason for changing this then it is because some other code is in the wrong place somehow. Possibly that's because some case isn't being handled in `invert` when it should be.

Reply:
> To me it looks like if there is a reason for changing this then it is because some other code is in the wrong place somehow. Possibly that's because some case isn't being handled in invert when it should be.

I think for `_is_lambert` the exception remains since, `invert` doesn't give a solution for this and I think it's pretty ineffective also ( as discussed in https://github.com/sympy/sympy/pull/21025#discussion_r589443553 ).